records/2024-04-11.2FA-magic-link.md

```diff
@@ -0,0 +1,170 @@
+
+#### Mitigations
+* Provide guidance to inform users of the risks of magic links, and steps they can take to mitigate them.
+* See the [Fallback Mitigation section](#Fallback-Mitigation---An-over-engineered-"last-resort"-solution)
+
+
+## Fallback Mitigation - An over-engineered "last resort" solution
+
+A fallback mitigation layer to protect against scenarios that are out of our control would be to establish a process for generating a unique visitor fingerprint. This fingerprint could be ephemerally stored for the duration of a magic link's validity period. That fingerprint could then be validated once the user clicks the magic link, and before beginning the final steps of the login process.
+
+The fingerprint could use the user's browser attributes, IP address, HTTPS session keys, etc to form a composite key that could then be hashed to uniquely identify the user. Since the success of the attacker is time dependent, [Argon2d](https://github.com/P-H-C/phc-winner-argon2) could be a good candidate for such a hashing implementation as it uses data-depending memory access to make it resistant to GPU based brute force attacks.
+
+```mermaid
+sequenceDiagram
+    actor User
+    participant Notify
+
+    User->>+Notify: Logs in
+        create participant User's browser
+    Notify->>+User's browser: Query browser attributes
+    User's browser-->>-Notify: Serve requested attributes
+
+    Notify->>Notify: Compose attributes into fingerprint
+        create participant Argon2d
+    Notify->>+Argon2d: Request hash of the fingerprint
+    Argon2d-->>-Notify: Hashed fingerprint
+        create participant DB
+    Notify->>+DB: Store fingerprint in the database as a login event
+    DB-->>-Notify: Fingerprint stored
+    Notify->>+User's browser: Respond with request for user to check their email
+    alt User never clicks the magic link
+        loop Celery task every X interval
+            Notify->>DB: Delete expired / stale fingerprint
+        end
+    end
+    alt User clicks magic link
+        User->>User's browser: Check email & Click magic link
+        User's browser-->>-Notify: Send login request
+        Notify->>User's browser: Re-request browser attributes
+        User's browser->>Notify: Requested attributes
+        Notify->>DB: Fetch Hash
+        DB-->>Notify: Hash
+        Notify->>Argon2d: Validate fingerprint hash
+        Argon2d-->>Notify: isValid
+
+        alt Invalid hash
+            Notify->>User's browser: Login failed
+        end
+        alt Valid hash
+            Notify->>User's browser: Log user in
+            Notify->>DB: Delete login event / fingerprint
+        end
+    end
+
+```
+
+Such an approach could allow us to regain control over some scenarios that typically arise from user action and behaviour that is out of our control:
```

Comment:
Thanks for explaining this scenario. The cost consideration is something that we need to take into account. What is industry standard here btw? Magic Links are becoming more common, do most folk implement this in the way suggested above?

Is there a link to the threat modeling exercise that was done for this ADR and if so can it be added here? What is the expectation from our security team on what level of defensiveness we need to implement?

Reply:
Here is the threat modeling doc: https://docs.google.com/document/d/1sIdrJ_dlkzwQoEVss0SMXjhdrLPYUenFlf3Xm8-amUc/edit

Reply:
I added a link to the thread modelling exercise doc under the `Security Considerations` section.